### PR TITLE
Further work on CMake build system, fix Visual Studio 2010

### PR DESCRIPTION
**Pull request by uboot**
This is basically an extension of #1. The additions to #1:
- Builds with Qt4 and VTK6
- Builds IDS camera support
- Builds with OpenCV 2 (in addition to version 3)
- Fixes build with Visual Studio 2010.

To my knowledge this is one of very few serious open source implementations of structured light scanning and I hope I can find time to decouple the implementation of the scanning code from the presentation layer (=Qt + VTK). As I commented earlier I think porting to CMake makes sense for this task.

### CODE REVIEW

**Comment by jakobwilm** [comment]
I'd like to use many of the changes and functionality added. However, the project is qmake based and I don't see compelling reason to change that. Any good arguments?

**Comment by hcw70** [comment]
The major argument is you are mentioning concrete versions of additional libraries like opencv, vtk et al. When i compiled your version, i needed to change some of those version specifiers to match my installation (Ubuntu 18.04).

With CMake you can (more or less) issue a "FindLibrary" and specify the minimum required library to use. So your build definitions become more portable, since most people wont need to fiddle within your qmakefiles any more since the version adaption happens automatically.

QtCreator can cope also with CMake based builds.

CMake is also well-documented, widely used and (most of the time) works flawesly...

**Comment by jakobwilm** [comment]
Well, broader library support and more people knowing about CMake speak for it. However, it has some downsides, namely much worse integration into QtCreator and a very convoluted syntax. Looking at the different CMakeList.txt stuff in the PRs, it also seems to me that maintainability suffers. 
From personal experience, I also rather change a few lines of a .pro file than debug a failing CMake build based on cryptic output.
So what it comes down to to me is whether there is a critical usecase which qmake simply cannot cover.

/Jakob

**Comment by hcw70** [comment]
@jakobwilm The critical use case here is to gain more easy compilation in different environments (Win32 / WIn64, Ubunutu 14, Ubuntu 16, Ubuntu 18, MacOS, ...) which mainly differ in the versions of required additional libraries and the presence of optional libraries.

A good build system should figure out these things by itself.

So most non-expert users are not able to fix these kind of errors by himself and are distracted from your project.

If I download a project to try it, and compile it according to documentation, and it does not compile,  i tend to delete it immediately if the most obvious fixes to the build dont help. And i am a "expert" user so to say, but the energy to put in to evaluate a SW whether it is worth to invest more energy should be as low as possible.

So if you want to enlarge the community of users of your SW (and i think you should, since it may become a good tool) you should at least:
- Provide a build system which allows flexiblility in the environments you support
- Listen to your users by letting them provide valuable feedback and merge requests
- Honestly react on reported issues 

I personally would like to use SLStudio since i would scan some stuff and use that in blender (faces, nature objects etc). 
What currently blocks me is:
- QMake bases build fails under Ubunutu 18 since some libs were not in the "correct" version. I was able to resolve them by modifying the qmake file.
- No supported camera i can use currently. 
-- I would use Webcams or a CHDK driven Canon Ixus, both are unsupported. 
-- I am already thinking about implementing QCamera support, which will open access to most system supported cameras IMHO on all QT supported plattforms (Win, Mac, Linux)
 -- Video4Linux support looked good, but does not seem to work completely.

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+add_subdirectory(${PROJECT_SOURCE_DIR}/src)
+add_subdirectory(${PROJECT_SOURCE_DIR}/src/camera)
+add_subdirectory(${PROJECT_SOURCE_DIR}/src/projector)
+
+#generate cmake config header file
+configure_file(${PROJECT_SOURCE_DIR}/slconfig.h.in
+        ${PROJECT_BINARY_DIR}/slconfig.h)
+
+#Set source file set
+file(GLOB UI_FILES src/SLStudio.ui
```
Comment (hcw70) [inline CMakeLists.txt]: File glob with listing all the files IMHO does not make much sense. It is for collecting a set of files based on a pattern, so you can i.e.
`file(GLOB UI_FILES src/*.ui)
`
or say explicitly say:
```
set(UI_FILES src/SLStudio.ui 
   src/SLPreferenceDialog.ui
   ... 
)
```

**Comment by uboot** [comment]
> The other changes in the PR i cant comment since i dunno the reaon (i.e. why some apis were changed from QSting -> std::string)...

This was a while ago but I am pretty sure that the original idea was to make the algorithmic core components (slalgorithm?) independent of Qt. As it seems that only minor changes were necessary to achieve this I think this still makes sense.

**Pull request closed without merging**